### PR TITLE
Adding a new testing metric reporter that can be used in unit testing…

### PR DESCRIPTION
**Pull request by buchi-busireddy**
… metrics logic.

Also cleaned up the code and the way metrics configs are passed around.

### CODE REVIEW

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java [hunk 1]
```diff
@@ -118,18 +135,49 @@ public String get(String key) {
...
+
+    int reportIntervalSec = DEFAULT_METRIC_REPORT_INTERVAL_SEC;
+    if (config.hasPath(METRICS_REPORT_INTERVAL_CONFIG_KEY)) {
+      reportIntervalSec = config.getInt(METRICS_REPORT_INTERVAL_CONFIG_KEY);
+    }
+
+    Map<String, String> defaultTags = new HashMap<>();
+
+    // Add the service name and other given tags to the default tags list.
+    defaultTags.put("app", serviceName);
```
Comment (surajpuvvada) [inline platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:169]: what happens when there are multiple instances of a service ?

Comment (buchi-busireddy) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:169]: They all get the same `app` tag, which is what we want. The instance details are added at the pod level if it's running in k8s. We could add the host name or IP also if needed later.

Comment (surajpuvvada) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:169]: Forgot that scrape config  will determine those additional labels like pod name that will get injected

**Comment by tim-mwangi** [comment]
@buchi-busireddy this significantly decreases codecov

**Comment by buchi-busireddy** [comment]
> @buchi-busireddy this significantly decreases codecov

@tim-mwangi Do you know why the ConfigUtilsTest is considered untracked by codecov? I think that's what is causing this reduction in coverage number.

Screen Shot 2020-08-12 at 9 40 14 PM

**Comment by buchi-busireddy** [comment]
@tim-mwangi Also codecov run has failed. I'm not sure if if the drop is really accurately. I'm merging this now since @samarth-gupta-traceable is blocked on this change but will get to see the coverage again and will fix if needed.

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java (hunk 1 above)
Comment (avinashkolluru) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:169]: Sorry, what is this scrape config? Are the pod/node etc tags automatically added to each metric somehow?

Comment (buchi-busireddy) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:169]: @avinashkolluru the way metrics are sent to Prom is this:

1. service exposes an endpoint `/metrics` where metrics are spit in Prom compatible fashion. No tags until now but with micro-meter, service itself will be adding some additional tags.
2. Prom collector calls the `/metrics` endpoint of each pod periodically (where is this configured??) and adds the labels of the pod to all those metrics before pushing to Prometheus. This is how we have `app` label on all metrics as of today, though the service isn't setting any labels.

With the new library, service code also can add labels, like `tenantid`

**Pull request merged**